File: backend/src/main/java/com/medcore/features/appointment/service/AppointmentService.java
package com.medcore.features.appointment.service;

import com.medcore.common.response.ApiResponse;
import com.medcore.common.response.PageResponse;
import com.medcore.features.appointment.dto.request.CreateAppointmentRequest;
import com.medcore.features.appointment.dto.request.UpdateAppointmentStatusRequest;
import com.medcore.features.appointment.dto.response.AppointmentResponse;

public interface AppointmentService {

    ApiResponse<AppointmentResponse> createAppointment(
            CreateAppointmentRequest request
    );
    
    ApiResponse<PageResponse<AppointmentResponse>> getAllAppointments(
            int page,
            int size,
            String sortBy,
            String sortDir
    );

    ApiResponse<AppointmentResponse> getAppointmentById(
            Long appointmentId
    );
    
    ApiResponse<PageResponse<AppointmentResponse>> getDoctorAppointments(
            Long doctorId,
            int page,
            int size
    );

    ApiResponse<PageResponse<AppointmentResponse>> getPatientAppointments(
            Long patientId,
            int page,
            int size
    );
    
    ApiResponse<AppointmentResponse> updateAppointmentStatus(
            Long appointmentId,
            UpdateAppointmentStatusRequest request
    );

    ApiResponse<String> cancelAppointment(Long appointmentId);
    
    ApiResponse<String> deleteAppointment(Long appointmentId);

    ApiResponse<String> restoreAppointment(Long appointmentId);
    
    ApiResponse<AppointmentResponse> checkInAppointment(
            Long appointmentId
    );
    
    ApiResponse<PageResponse<AppointmentResponse>> getTodayAppointments(
            int page,
            int size,
            String sortBy,
            String sortDir
    );
}

File: frontend/src/services/appointment.service.js
import { apiFetch } from "@/lib/api-client";

async function handleResponse(response) {
  const result = await response.json();

  if (!response.ok || !result.success) {
    throw new Error(result.message || "Something went wrong");
  }

  return result;
}

 

export async function createAppointment(data) {
  const response = await apiFetch("/appointments", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  return handleResponse(response);
}

 
export async function getMyAppointments({
  page = 0,
  size = 20,
  sortBy = "appointmentDate",
  sortDir = "desc",
} = {}) {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
    sortBy,
    sortDir,
  });

  const response = await apiFetch(`/appointments/me?${params.toString()}`);
  return handleResponse(response);
}


export async function getAppointments({
  page = 0,
  size = 10,
  sortBy = "appointmentDate",
  sortDir = "asc",
} = {}) {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
    sortBy,
    sortDir,
  });

  const response = await apiFetch(
    `/appointments?${params.toString()}`
  );

  return handleResponse(response);
}

 

export async function getTodayAppointments({
  page = 0,
  size = 10,
  sortBy = "startTime",
  sortDir = "asc",
} = {}) {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
    sortBy,
    sortDir,
  });

  const response = await apiFetch(
    `/appointments/today?${params.toString()}`
  );

  return handleResponse(response);
}

 

export async function getDoctorAppointments(
  doctorId,
  {
    page = 0,
    size = 10,
  } = {}
) {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
  });

  const response = await apiFetch(
    `/appointments/doctor/${doctorId}?${params.toString()}`
  );

  return handleResponse(response);
}

 

export async function getPatientAppointments(
  patientId,
  {
    page = 0,
    size = 10,
  } = {}
) {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
  });

  const response = await apiFetch(
    `/appointments/patient/${patientId}?${params.toString()}`
  );

  return handleResponse(response);
}

 

export async function getAppointmentById(appointmentId) {
  const response = await apiFetch(
    `/appointments/${appointmentId}`
  );

  return handleResponse(response);
}

 

export async function updateAppointmentStatus(
  appointmentId,
  status
) {
  const response = await apiFetch(
    `/appointments/${appointmentId}/status`,
    {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        status,
      }),
    }
  );

  return handleResponse(response);
}
 

export async function checkInAppointment(appointmentId) {
  const response = await apiFetch(
    `/appointments/${appointmentId}/check-in`,
    {
      method: "PATCH",
    }
  );

  return handleResponse(response);
}



export async function cancelAppointment(appointmentId) {
  const response = await apiFetch(
    `/appointments/${appointmentId}/cancel`,
    {
      method: "PATCH",
    }
  );

  return handleResponse(response);
}

 

export async function deleteAppointment(appointmentId) {
  const response = await apiFetch(
    `/appointments/${appointmentId}`,
    {
      method: "DELETE",
    }
  );

  return handleResponse(response);
}

 

export async function restoreAppointment(appointmentId) {
  const response = await apiFetch(
    `/appointments/${appointmentId}/restore`,
    {
      method: "PATCH",
    }
  );

  return handleResponse(response);
}

 

export async function getAvailableSlots(
  doctorId,
  appointmentDate
) {
  const params = new URLSearchParams({
    doctorId: String(doctorId),
    appointmentDate,
  });

  const response = await apiFetch(
    `/appointments/available-slots?${params.toString()}`
  );

  return handleResponse(response);
}